File: solutions/python/robot-simulator/1/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
EAST = 4
NORTH = 1
WEST = 2
SOUTH = 3


class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.coordinates=(x_pos,y_pos)
        self.direction=direction

    def move(self,motion):
        for char in motion:
            if char=='A':
                if self.direction==NORTH:
                    self.coordinates=(self.coordinates[0],self.coordinates[1]+1)
                elif self.direction==SOUTH:
                    self.coordinates=(self.coordinates[0],self.coordinates[1]-1)
                elif self.direction==EAST:
                    self.coordinates=(self.coordinates[0]+1,self.coordinates[1])
                elif self.direction==WEST:
                    self.coordinates=(self.coordinates[0]-1,self.coordinates[1])
            elif char=='R':
                if self.direction==NORTH:
                    self.direction=EAST
                elif self.direction==SOUTH:
                    self.direction=WEST
                elif self.direction==EAST:
                    self.direction=SOUTH
                elif self.direction==WEST:
                    self.direction=NORTH
            elif char=='L':
                if self.direction==NORTH:
                    self.direction=WEST
                elif self.direction==SOUTH:
                    self.direction=EAST
                elif self.direction==EAST:
                    self.direction=NORTH
                elif self.direction==WEST:
                    self.direction=SOUTH

File: solutions/typescript/robot-simulator/1/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

export class Robot {
  private direction: Direction
  private position: Coordinates
  
  constructor (direction:Direction='north', x: number=0, y: number=0){
    this.direction=direction
    this.position=[x,y]
  }
  get bearing(): Direction {
    return this.direction
  }

  get coordinates(): Coordinates {
    return this.position
  }

  place({direction,x,y}: { x: number; y: number; direction: Direction }) {
    if (['north', 'east', 'south', 'west'].indexOf(direction) < 0) {
      throw new InvalidInputError("Not allowed direction");
    }
    this.direction=direction
    this.position=[x,y]
  }

  evaluate(instructions: string) {
    let i: number = 0
    while (i < instructions.length) {
      if (instructions.charAt(i) == 'A') {
        switch (this.direction) {
          case 'north': {
            this.position[1]=this.position[1]+1
            break
          }
          case 'east': {
            this.position[0]=this.position[0]+1
            break
          }
          case 'south': {
            this.position[1]=this.position[1]-1
            break
          }
          case 'west': {
            this.position[0]=this.position[0]-1
            break
          }
        }
      }
      else if (instructions.charAt(i) == 'R') {
        switch (this.direction) {
          case 'north': {
            this.direction='east'
            break
          }
          case 'east': {
            this.direction='south'
            break
          }
          case 'south': {
            this.direction='west'
            break
          }
          case 'west': {
            this.direction='north'
            break
          }
        }
      }
      else if (instructions.charAt(i) == 'L') {
        switch (this.direction) {
          case 'north': {
            this.direction='west'
            break
          }
          case 'east': {
            this.direction='north'
            break
          }
          case 'south': {
            this.direction='east'
            break
          }
          case 'west': {
            this.direction='south'
            break
          }
        }
      }
      i++;
    }

  }
}
